les/utils/limiter.go

```diff
@@ -0,0 +1,405 @@
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package utils
+
+import (
+	"sort"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/p2p/enode"
+)
+
+// Limiter protects a network request serving mechanism from denial-of-service attacks.
+// It limits the total amount of resources used for serving requests while ensuring that
+// the most valuable connections always have a reasonable chance of being served.
+type Limiter struct {
+	lock sync.Mutex
+	cond *sync.Cond
+	quit bool
+
+	nodes                                             map[enode.ID]*nodeQueue
+	addresses                                         map[string]*addressGroup
+	addressSelect, valueSelect                        *WeightedRandomSelect
+	maxValue                                          float64
+	maxPriorWeight, totalPriorWeight, totalPriorLimit uint64
+	lastValueSelected                                 bool
+}
+
+// nodeQueue represents queued requests coming from a single node ID
+type nodeQueue struct {
+	queue                                     []request
+	id                                        enode.ID
+	address                                   string
+	value                                     float64
+	flatWeight, valueWeight, totalPriorWeight uint64
+	groupIndex                                int
+	relCost                                   float64
+}
+
+// addressGroup is a group of node IDs that have sent their last requests from the same
+// network address
+type addressGroup struct {
+	nodes                      []*nodeQueue
+	nodeSelect                 *WeightedRandomSelect
+	sumFlatWeight, groupWeight uint64
+}
+
+// request represents an incoming request scheduled for processing
+type request struct {
+	process     chan chan float64
+	priorWeight uint64
+}
+
+// flatWeight distributes weights equally between each active network adress
+func flatWeight(item interface{}) uint64 { return item.(*nodeQueue).flatWeight }
+
+// add adds the node queue to the address group. It is the caller's responsibility to
+// add the address group to the address map and the address selector if it wasn't
+// there before.
+func (ag *addressGroup) add(nq *nodeQueue) {
+	if nq.groupIndex != -1 {
+		panic("added node queue is already in an address group")
+	}
+	l := len(ag.nodes)
+	if l == 1 {
```

Comment:
I don't understand why we need to specialize the single element set?

If there is a single element in the selection set, we can still build the set for it

Reply:
Done. The original version was meant to be a performance optimization but I tried and it does not make a big difference (less than 10%).